File: typescript/src/adapters/anthropic.ts
import type { Canopy } from "../client.js";

export interface AnthropicTool {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
}

export interface AnthropicContentBlock {
  type: string;
  id?: string;
  name?: string;
  input?: Record<string, unknown>;
}

export interface AnthropicToolResultBlock {
  type: "tool_result";
  tool_use_id: string;
  content: string;
}

export class AnthropicAdapter {
  constructor(private readonly canopy: Canopy) {}

  /**
   * Canopy's canonical tools shaped for Anthropic's Messages API —
   * `[{ name, description, input_schema }]`. Pass directly as `tools` on
   * `messages.create`.
   */
  tools(): AnthropicTool[] {
    return this.canopy.getTools().map((t) => ({
      name: t.name,
      description: t.description,
      input_schema: t.parameters,
    }));
  }

  /**
   * Execute every Canopy `tool_use` block from an assistant message and
   * return Anthropic-shaped `tool_result` blocks. Wrap them in a user
   * message (`{ role: "user", content: <result-blocks> }`) and append for
   * the next turn.
   *
   * Non-`tool_use` blocks and blocks naming a non-Canopy tool are skipped.
   * Errors thrown by the underlying Canopy method are embedded as `{ error }`
   * JSON in the result content so the LLM can react.
   */
  async dispatch(
    content: AnthropicContentBlock[] | null | undefined,
  ): Promise<AnthropicToolResultBlock[]> {
    if (!content?.length) return [];
    const byName = new Map(this.canopy.getTools().map((t) => [t.name, t]));
    const out: AnthropicToolResultBlock[] = [];
    for (const block of content) {
      if (block.type !== "tool_use" || !block.id || !block.name) continue;
      const tool = byName.get(block.name);
      if (!tool) continue;
      let resultContent: string;
      try {
        const result = await tool.execute(block.input ?? {});
        resultContent = JSON.stringify(result ?? null);
      } catch (err) {
        resultContent = JSON.stringify({
          error: err instanceof Error ? err.message : String(err),
        });
      }
      out.push({
        type: "tool_result",
        tool_use_id: block.id,
        content: resultContent,
      });
    }
    return out;
  }
}

File: python/canopy_ai/adapters/anthropic.py
"""Anthropic Messages adapter.

Returns plain ``dict``s; no ``anthropic`` package import. ``tools()`` matches
``messages.create(tools=...)`` shape; ``dispatch()`` consumes assistant
content blocks and returns ``tool_result`` blocks the caller wraps in a user
message.
"""

from __future__ import annotations

import inspect
import json
from typing import TYPE_CHECKING, Any, Iterable

if TYPE_CHECKING:
    from canopy_ai.async_client import AsyncCanopy
    from canopy_ai.client import Canopy


def _tools_for(canopy: Canopy | AsyncCanopy) -> list[dict[str, Any]]:
    raw = canopy.get_tools()
    return [
        {
            "name": t["name"],
            "description": t["description"],
            "input_schema": t["parameters"],
        }
        for t in raw
    ]


def _executors_by_name(canopy: Canopy | AsyncCanopy) -> dict[str, Any]:
    return {t["name"]: t["execute"] for t in canopy.get_tools()}


def _read_block(block: Any) -> tuple[str, str, dict[str, Any]] | None:
    btype = getattr(block, "type", None) or (
        block.get("type") if isinstance(block, dict) else None
    )
    if btype != "tool_use":
        return None
    block_id = getattr(block, "id", None) or (
        block.get("id") if isinstance(block, dict) else None
    )
    name = getattr(block, "name", None) or (
        block.get("name") if isinstance(block, dict) else None
    )
    raw_input = getattr(block, "input", None)
    if raw_input is None and isinstance(block, dict):
        raw_input = block.get("input")
    if not block_id or not name:
        return None
    if not isinstance(raw_input, dict):
        raw_input = {}
    return (str(block_id), str(name), raw_input)


def _embed_error(err: Exception) -> str:
    return json.dumps({"error": str(err)})


class AnthropicAdapter:
    """Sync Anthropic adapter for :class:`canopy_ai.Canopy`."""

    def __init__(self, canopy: "Canopy") -> None:
        self._canopy = canopy

    def tools(self) -> list[dict[str, Any]]:
        """Canopy tools shaped for ``messages.create(tools=...)``."""
        return _tools_for(self._canopy)

    def dispatch(self, content: Iterable[Any] | None) -> list[dict[str, Any]]:
        """Execute every ``tool_use`` block and return ``tool_result`` blocks.

        Wrap the returned list in ``{"role": "user", "content": <blocks>}``
        and append for the next ``messages.create()`` turn.
        """
        if not content:
            return []
        executors = _executors_by_name(self._canopy)
        out: list[dict[str, Any]] = []
        for block in content:
            parsed = _read_block(block)
            if parsed is None:
                continue
            tu_id, name, args = parsed
            execute = executors.get(name)
            if execute is None:
                continue
            try:
                result = execute(args)
                result_content = json.dumps(result)
            except Exception as err:  # noqa: BLE001
                result_content = _embed_error(err)
            out.append(
                {
                    "type": "tool_result",
                    "tool_use_id": tu_id,
                    "content": result_content,
                }
            )
        return out


class AsyncAnthropicAdapter:
    """Async Anthropic adapter for :class:`canopy_ai.AsyncCanopy`."""

    def __init__(self, canopy: "AsyncCanopy") -> None:
        self._canopy = canopy

    def tools(self) -> list[dict[str, Any]]:
        return _tools_for(self._canopy)

    async def dispatch(
        self, content: Iterable[Any] | None
    ) -> list[dict[str, Any]]:
        if not content:
            return []
        executors = _executors_by_name(self._canopy)
        out: list[dict[str, Any]] = []
        for block in content:
            parsed = _read_block(block)
            if parsed is None:
                continue
            tu_id, name, args = parsed
            execute = executors.get(name)
            if execute is None:
                continue
            try:
                result = execute(args)
                if inspect.isawaitable(result):
                    result = await result
                result_content = json.dumps(result)
            except Exception as err:  # noqa: BLE001
                result_content = _embed_error(err)
            out.append(
                {
                    "type": "tool_result",
                    "tool_use_id": tu_id,
                    "content": result_content,
                }
            )
        return out
